tnfsd -b <address>: listen on one address instead of all of them.

fly.io delivers a UDP service's packets only to a program listening on the machine's fly-global-services
address; replies from a socket bound to all addresses leave from the wrong one and never arrive. The
Dockerfile applies this to tnfsd v26.1015.1 (4aab910) from github.com/FujiNetWIFI/tnfsd. Small and general
enough to offer upstream, after which this file can go.

diff --git a/src/datagram.c b/src/datagram.c
index 3d43fef..89ddb58 100644
--- a/src/datagram.c
+++ b/src/datagram.c
@@ -38,10 +38,12 @@ TNFS daemon datagram handler
 #include <sys/socket.h>
 #include <netinet/in.h>
 #include <netinet/tcp.h>
+#include <netdb.h>
 #endif
 
 #ifdef WIN32
 #include <winsock2.h>
+#include <ws2tcpip.h>
 #include <windows.h>
 #endif
 
@@ -58,6 +60,7 @@ TNFS daemon datagram handler
 
 int sockfd;			/* UDP global socket file descriptor */
 int tcplistenfd;	/* TCP listening socket file descriptor */
+const char *tnfs_bindaddr = NULL;	/* address to listen on (-b); NULL = all */
 
 /* Set by tnfsd_stop(). The loop used to exit only because a signal made
  * tnfs_event_wait() fail, which is indistinguishable from a real error. */
@@ -151,9 +154,34 @@ const char *get_cmd_name(uint8_t cmd)
 	return "UNKNOWN_CMD";
 }
 
+/* The IPv4 address to listen on: tnfs_bindaddr (a name or a dotted quad), or
+ * all addresses when it is NULL. Returns -1 if the name doesn't resolve. */
+static int tnfs_listenaddr(struct in_addr *addr)
+{
+	struct addrinfo hints, *res;
+
+	if (tnfs_bindaddr == NULL)
+	{
+		addr->s_addr = htonl(INADDR_ANY);
+		return 0;
+	}
+	memset(&hints, 0, sizeof(hints));
+	hints.ai_family = AF_INET;
+	if (getaddrinfo(tnfs_bindaddr, NULL, &hints, &res) != 0 || res == NULL)
+	{
+		LOG("Can't resolve the bind address %s\n", tnfs_bindaddr);
+		return -1;
+	}
+	*addr = ((struct sockaddr_in *)res->ai_addr)->sin_addr;
+	freeaddrinfo(res);
+	LOG("Listening on %s only\n", tnfs_bindaddr);
+	return 0;
+}
+
 int tnfs_sockinit(int port)
 {
 	struct sockaddr_in servaddr;
+	struct in_addr listenaddr;
 
 #ifdef WIN32
 	WSADATA wsaData;
@@ -164,6 +192,9 @@ int tnfs_sockinit(int port)
 	}
 #endif
 
+	if (tnfs_listenaddr(&listenaddr) < 0)
+		return -1;
+
 	/* Create the UDP socket */
 	sockfd = socket(AF_INET, SOCK_DGRAM, 0);
 	if (sockfd < 0)
@@ -175,7 +206,7 @@ int tnfs_sockinit(int port)
 	/* set up the network */
 	memset(&servaddr, 0, sizeof(servaddr));
 	servaddr.sin_family = AF_INET;
-	servaddr.sin_addr.s_addr = htons(INADDR_ANY);
+	servaddr.sin_addr = listenaddr;
 	servaddr.sin_port = htons(port);
 
 	if (bind(sockfd, (struct sockaddr *)&servaddr, sizeof(servaddr)) < 0)
@@ -231,7 +262,7 @@ int tnfs_sockinit(int port)
 
 	memset(&servaddr, 0, sizeof(servaddr));
 	servaddr.sin_family = AF_INET;
-	servaddr.sin_addr.s_addr = htons(INADDR_ANY);
+	servaddr.sin_addr = listenaddr;
 	servaddr.sin_port = htons(port);
 	if (bind(tcplistenfd, (struct sockaddr *)&servaddr,
 			 sizeof(servaddr)) < 0)
diff --git a/src/datagram.h b/src/datagram.h
index efb8e17..f26ca03 100644
--- a/src/datagram.h
+++ b/src/datagram.h
@@ -48,6 +48,7 @@
 extern volatile sig_atomic_t tnfs_stop_requested;
 
 /* Handle the socket interface */
+extern const char *tnfs_bindaddr;	/* address to listen on; NULL = all */
 int tnfs_sockinit(int port);
 void tnfs_sockclose();
 void tnfs_mainloop();
diff --git a/src/main.c b/src/main.c
index 8a8661c..d304a62 100644
--- a/src/main.c
+++ b/src/main.c
@@ -34,6 +34,7 @@
 #include "chroot.h"
 #include "log.h"
 #include "tnfsd.h"
+#include "datagram.h"
 
 /* declare the main() - it won't be used elsewhere so I'll not bother
  * with putting it in a .h file */
@@ -54,9 +55,9 @@ int main(int argc, char **argv)
     char *root_path = NULL;
 
     #ifdef ENABLE_CHROOT
-    while((opt = getopt(argc, argv, "aru:g:p:")) != -1)
+    while((opt = getopt(argc, argv, "arb:u:g:p:")) != -1)
     #else
-    while((opt = getopt(argc, argv, "arp:")) != -1)
+    while((opt = getopt(argc, argv, "arb:p:")) != -1)
     #endif
     {
         switch(opt)
@@ -70,6 +71,9 @@ int main(int argc, char **argv)
             case 'a':
                 atari_mode = true;
                 break;
+            case 'b':
+                tnfs_bindaddr = optarg;
+                break;
             #ifdef ENABLE_CHROOT
             case 'u':
                 uvalue = optarg;
@@ -152,10 +156,12 @@ int main(int argc, char **argv)
 void print_usage()
 {
     #ifdef ENABLE_CHROOT
-    fprintf(stderr, "Usage: tnfsd [-u <username> -g <group> -p <port> -r -a] <root dir>\n");
+    fprintf(stderr, "Usage: tnfsd [-u <username> -g <group> -p <port> -b <address> -r -a] <root dir>\n");
     fprintf(stderr, "  -a  Enable Atari mode (present binary files as ATR images)\n");
+    fprintf(stderr, "  -b  Listen on this address only (a name or an IPv4 address); default all\n");
     #else
-    fprintf(stderr, "Usage: tnfsd [-p <port> -r -a] <root dir>\n");
+    fprintf(stderr, "Usage: tnfsd [-p <port> -b <address> -r -a] <root dir>\n");
     fprintf(stderr, "  -a  Enable Atari mode (present binary files as ATR images)\n");
+    fprintf(stderr, "  -b  Listen on this address only (a name or an IPv4 address); default all\n");
     #endif
 }
\ No newline at end of file
